### ISSUE TITLE
Empty key usage extension

### ISSUE THREAD

**Issue by dulanshuangqiao** [openssl/openssl#25937]
I used the generated keyusage extension certificate with an empty value to perform a format conversion experiment.
The conversion command of openssl successfully achieved the format conversion, but the converter of golang considered this extension invalid, so I used openssl x509 -text -in fd.der -noout to view the extension, and got different results from using certtool -i --inraw --infile fd.der.
As shown in the figure: openssl parses the keyusage extension without a value, and the result content is empty, while gnutls chooses to report an error.
According to the description of keyusage in RFC5280, when the keyusage extension appears in a certificate, at least one bit must be set to 1. Keyusage without a value should be an error.
![gnutls](https://github.com/user-attachments/assets/647ca92c-7d74-43b1-a45e-529ee7a740b8)
![openssl](https://github.com/user-attachments/assets/5c8b5545-8863-4bfe-a3f7-f96364684ac1)
[case.zip](https://github.com/user-attachments/files/17725377/case.zip)
（I provided the certificate used.）

**Comment by dcooper16** [comment]
While RFC 5280 states that "at least one of the bits MUST be set to 1," X.509 does not. X.509 just says that if the extension is present and all bits are zero, then the key is intended for a purpose other than the ones listed.

**Comment by t8m** [comment]
Yeah - we would need something like an optional strict parsing/verification of certs that would be used for example in the TLS context but not in general.

**Comment by dulanshuangqiao** [comment]
> While RFC 5280 states that "at least one of the bits MUST be set to 1," X.509 does not. X.509 just says that if the extension is present and all bits are zero, then the key is intended for a purpose other than the ones listed.

Do you mean the X.509 rules like RFC5280, or the X.509 commands of OpenSSL?
Because GnutLS strictly follows RFC5280, it thinks this is invalid and throws an error.

**Comment by dcooper16** [comment]
> > While RFC 5280 states that "at least one of the bits MUST be set to 1," X.509 does not. X.509 just says that if the extension is present and all bits are zero, then the key is intended for a purpose other than the ones listed.
> 
> Do you mean the X.509 rules like RFC5280, or the X.509 commands of OpenSSL? Because GnutLS strictly follows RFC5280, it thinks this is invalid and throws an error.

By X.509 I mean the actual ITU-T standard (https://www.itu.int/rec/T-REC-X.509/en). X.509 is the base standard, and RFC 5280 is an IETF profile of that standard. So, RFC 5280 includes recommendations/requirements that are not imposed by the base standard (X.509). For example, X.509 says that the nameConstraints extension may be flagged as either critical or non-critical, but RFC 5280 says that conforming CAs must mark it as critical. However, Section 6.1 of RFC 5280 notes that the path validation algorithm in that document is designed to determine whether a certificate is valid according to the X.509 path validation rules, not to determine whether the certificate was issued in conformance with the RFC 5280 profile.

For path validation, OpenSSL has an `-x509_strict` option for which "errors are thrown on certificates not compliant with RFC 5280." However, this comment is about printing a certificate using `openssl x509 -text`, and I don't know whether that code was intended to perform any checks on the certificate (whether for conformance to X.509 or RFC 5280).

**Comment by dcooper16** [comment]
I didn't look at the sample certificates, but you are correct that the encoding in Figure 1 is incorrect. The OCTET STRING for the extension value must include a DER-encoded value, it can not be empty. In the case of keyUsage, it must be the DER encoding of a BIT STRING. X.509 allows this BIT STRING to be all zeros, as in Figure 2, but it must be present.

The issue you seem to be asking, however, is whether `openssl x509 -text` should throw an error if there is an encoding error in the certificate. I am not involved in the development of OpenSSL, so I can't answer that. My guess, however, is that the function for pretty-printing the contents of a certificate was not designed for check for and warn about encoding errors.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
